subprojects/core/src/main/java/org/gradle/api/reflect/TypeOf.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.reflect;
+
+import com.google.common.base.Objects;
+import com.google.common.reflect.TypeToken;
```

Comment:
@adammurdoch, from what I can see, `ModelType<T>` doesn't provide a way to retrieve the original `java.lang.reflect.Type` from it and we were counting on that for the code generation that needs to happen on the gradle-script-kotlin side.

Since `TypeOf<T>` is only used for four operations currently - `getType(): java.lang.reflect.Type`, `equals(Object): boolean`, `isAssignableFrom(TypeOf<?>): boolean` and `getSimpleName(): String` - and there's no reason to believe the underlying `Class<?>` instances would need to be garbage collected, I'm thinking it would be best to move on with the current `TypeToken<T>` based implementation.

The alternative, it seems, would be to expose a lot more operations and basically replicate most of the `java.lang.reflect.Type` API.

What do you think?